src/auth.rs

```diff
@@ -3,41 +3,95 @@ use crate::log::Log;
 ) -> Result<bool> {
     let keys = keys_from_file(Path::new(keys_file_path))?;
+    // If no CA keys file is provided, then initialize ca_keys as an empty set.
+    let ca_keys_file_path = ca_keys_file_path.unwrap_or("");
+    let mut ca_keys = HashSet::new();
+    if !ca_keys_file_path.is_empty() {
+        ca_keys = keys_from_file(Path::new(ca_keys_file_path))?;
+    }
+
     for key in agent.list_identities()? {
-        if keys.contains(key.key_data()) {
-            log.debug(format!(
-                "found a matching key: {}",
-                key.fingerprint(Default::default())
-            ))?;
-            // Allow sign_and_verify() to return RemoteFailure (key not loaded / present),
-            // and try the next configured key
-            match sign_and_verify(&key, &mut agent) {
-                Ok(res) => return Ok(res),
-                Err(e) => {
-                    if let Some(SACError::RemoteFailure) = e.downcast_ref::<SACError>() {
-                        log.debug("SSHAgent: RemoteFailure; trying next key")?;
-                        continue;
-                    } else {
-                        return Err(e);
+        // A key can be a pubkey identity or a certificate identity.
+        // Adapt the logic to handle both cases.
+        match key {
+            Identity::PublicKey(key) => {
+                if keys.contains(&key) {
+                    log.debug(format!(
+                        "found a matching key: {}, comment: {}",
+                        key.fingerprint(Default::default()),
+                        key.comment()
+                    ))?;
+                    // Allow sign_and_verify() to return RemoteFailure (key not loaded / present),
+                    // and try the next configured key
+                    match sign_and_verify(&Identity::PublicKey(key), &mut agent) {
+                        Ok(res) => return Ok(res),
+                        Err(e) => {
+                            if let Some(SACError::RemoteFailure) = e.downcast_ref::<SACError>() {
+                                log.debug("SSHAgent: RemoteFailure; trying next key")?;
+                                continue;
+                            } else {
+                                return Err(e);
+                            }
+                        }
+                    }
+                }
+            }
+            Identity::Certificate(key) => {
+                for ca_key in ca_keys.iter() {
+                    if ca_key.key_data() == key.signature_key() {
+                        log.debug(format!(
+                            "found a matching certificate signed by trusted ca: {}, ca_key: {}",
+                            ca_key.comment(),
+                            ca_key.fingerprint(Default::default())
+                        ))?;
+                        // Validate the certificate against the CA key
+                        match key.validate([&ca_key.fingerprint(Default::default())]) {
```

Comment:
it seems only SHA256 fingerprints are valid at this time, so let's make that explict: https://github.com/RustCrypto/SSH/blob/master/ssh-key/src/certificate.rs#L372

Also, for when this is in a separate function, we should use validate_at so that we can verify that the time based invalidation indeed works as it should